pkg/apis/resource/types.go

```diff
@@ -37,64 +284,361 @@ type ResourceClaim struct {
+	//    device.attributes["dra.example.com"].driverVersion
+	//
+	// The type of each entry varies, depending on the attribute
+	// that is being looked up. The domain lookup returns an empty
+	// map if there is no attribute with that domain. However,
+	// unknown identifiers then trigger a runtime error.
+	//
+	// The `cel.bind` function is enabled and can be used to simplify
+	// expressions that access multiple attributes with the same domain:
+	//
+	//    cel.bind(dra, device.attributes["dra.example.com"], dra.someBool && dra.anotherBool)
+	//
+	// Capacities associated with a device are made available through a
+	// nested `device.capacities` map the same way as attributes.
+	//
+	// The `device.driverName` string variable can be used to check for a specific
+	// driver explicitly in a filter that is meant to work for devices from
+	// different vendors. It is provided by Kubernetes and matches the
+	// `driverName` from the ResourceSlice which provides the device.
+	//
+	// The CEL expression is applied to *all* available devices from any driver.
+	// The expression has to check for existence of an attribute when it is not
+	// certain that it is provided because runtime errors are not automatically
+	// treated as "don't select device". Instead, device selection fails completely
+	// and reports the error.
+	//
+	// Some more examples:
+	//
+	//    "memory" in device.capacities["dra.example.com"] && # Is the capacity available?
+	//       device.capacities["dra.example.com"].memory.isGreaterThan(quantity("1Gi")) # >= 1Gi
+	//
+	//    device.attributes["dra.example.com"].driverVersion.isGreaterThan(semver("1.0.0")) # >= v1.0.0, runtime error if not available
+	//
+	//    device.driverName == "dra.example.com" # any device from that driver
+	//
+	// +required
+	Expression string
+}
+
+// Exactly one field must be set.
+type Amount struct {
+	// All, if set, asks for all devices matching the selectors.
+	// Allocation fails if not all of them are available, unless admin
+	// access is requested. Admin access is granted also for
+	// devices which are in use.
+	//
+	// May only be set to true.
+	//
+	// +optional
+	All *bool
+
+	// A fixed number of devices matching the selectors.
+	//
+	// Must be greater than 0.
+	//
 	// +optional
+	ExactCount *int64
+}
+
+// Besides the request name slice, constraint must have exactly one field set.
+type Constraint struct {
+	// The constraint applies to devices in these requests. A single entry is okay
+	// and used when that request is for multiple devices.
+	//
+	// If empty, the constrain applies to all devices in the claim.
+	//
+	// +optional
+	// +listType=atomic
+	RequestNames []string
+
+	// The devices must have this attribute and its value must be the same.
+	//
+	// For example, if you specified "dra.example.com/numa" (a hypothetical example!),
+	// then only devices in the same NUMA node will be chosen.
+	//
+	// +required
+	MatchAttribute *string
+
+	// TODO (?)
+	// MatchQuantity *string
```

Comment:
"capacities" are not meant to be used for matching, as discussed in the meantime. Will remove...